### PR TITLE
Support NCCL v2.12

### PR DESCRIPTION
**Pull request by rashikakheria**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/nccl_ofi_net.c [hunk 1]
```diff
@@ -1180,6 +1180,19 @@ static ncclResult_t set_nic_props_default(int dev, struct fi_info *nic_prov,
...
+	 *
+	 * This value is chosen by measuring all reduce latency for
+	 * different NCCL algorithms and using that to calculate intra node
+	 * latency based on NCCL's tuning algorithm.
+	 *
+	 * A few different values around this value were tried to see which
+	 * chose the correct algorithm (tree or ring) most times across
+	 * different message and cluster sizes.
+	 */
+	props->latency = 150;
```
Comment (nzmsv) [inline src/nccl_ofi_net.c:1194]: Does using an EFA-specific value make sense here?  The master branch should be for generic for any OFI provider, no?

Comment (rashikakheria) [reply src/nccl_ofi_net.c:1194]: Yes, you are right! will remove it.

**Comment by rashikakheria** [comment]
@dmaryin FYI, this is the PR for supporting NCCL v2.12 (https://github.com/NVIDIA/nccl/tree/v2.12; would also need to apply this PR https://github.com/NVIDIA/nccl/pull/635). I have tested this against efa, tcp and sockets provider. Feel free to test it against psm3 provider.

**Comment by dmaryin** [comment]
> @dmaryin FYI, this is the PR for supporting NCCL v2.12 (https://github.com/NVIDIA/nccl/tree/v2.12; would also need to apply this PR [NVIDIA/nccl#635](https://github.com/NVIDIA/nccl/pull/635)). I have tested this against efa, tcp and sockets provider. Feel free to test it against psm3 provider.

Thanks a lot, got it

File: include/nccl_ofi.h [hunk 2]
```diff
@@ -195,6 +222,8 @@ typedef struct pending_reqs_q {
 typedef struct nccl_ofi_handle {
 	char ep_name[MAX_EP_ADDR];
 	uint64_t tag;
+	/* Save temporary communicator state when creating send communicator */
+	save_comm_state_t state;
```
Comment (nzmsv) [inline include/nccl_ofi.h:239]: Does this work with older versions of NCCL or will we have breakage similar to `aws` branch?

**Review by nzmsv — APPROVED** [review]

**Pull request merged**